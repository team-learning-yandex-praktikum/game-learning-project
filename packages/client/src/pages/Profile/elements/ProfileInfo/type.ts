import { UserDTO } from '@api/auth/types'

export type ProfileInfoProps = {
  user: UserDTO
  onSaveProfile: (data: Omit<UserDTO, 'id' | 'display_name' | 'avatar'>) => void
}
