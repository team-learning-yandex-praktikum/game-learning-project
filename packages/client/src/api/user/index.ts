import { BaseApi } from '@api/baseApi'
import { UserDTO } from '@api/auth/types'
import { UpdatingPasswordDTO, UpdatingProfileDTO } from './types'

export class UserApi extends BaseApi {
  constructor() {
    super('user')
  }

  updateProfile = async (profileData: UpdatingProfileDTO) => {
    const { data } = await this.client.put<UserDTO>('profile', profileData)
    return data
  }

  updateAvatar = async (avatarData: FormData) => {
    const { data } = await this.client.put<UserDTO>(
      'profile/avatar',
      avatarData
    )
    return data
  }

  updatePassword = async (passwordData: UpdatingPasswordDTO) => {
    const { data } = await this.client.put('password', passwordData)
    return data
  }

  getUser = async (id: string) => {
    const { data } = await this.client.get<UserDTO>(id)
    return data
  }
}

export const userApi = new UserApi()
