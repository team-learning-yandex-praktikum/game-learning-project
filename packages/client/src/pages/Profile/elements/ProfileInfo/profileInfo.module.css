.title {
  margin: 20px 0 40px;
  color: var(--color-accent-secondary);
  composes: font-base-xl from global;
  font-weight: 600;
}

.fields {
  display: flex;
  flex-direction: column;
  gap: 10px;
  margin-bottom: 40px;
}

.button {
  margin-bottom: 20px;
}
