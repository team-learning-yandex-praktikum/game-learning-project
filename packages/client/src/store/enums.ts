export enum LoadStatus {
    idle = 'idle',
    loading = 'loading',
    complete = 'complete',
    failed = 'failed',
}
