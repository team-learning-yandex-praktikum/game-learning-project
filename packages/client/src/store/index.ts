export { store } from './store'
export type { AppStore, AppDispatch, RootState } from './store'
