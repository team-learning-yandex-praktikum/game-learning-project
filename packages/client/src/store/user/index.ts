export { userSlice, userSelectors, userActions } from './userSlice'
export * from './thunks'
