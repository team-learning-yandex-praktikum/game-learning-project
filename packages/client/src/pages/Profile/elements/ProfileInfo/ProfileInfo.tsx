import { FC, MouseEvent, useCallback, useMemo, useState } from 'react'
import styles from './profileInfo.module.css'
import Title from '@components/Title'
import Avatar from '@pages/Profile/elements/Avatar'
import Form from '@components/Form'
import { fieldsConfig } from '@utils/validation/fieldsConfig'
import { omit, pick } from 'lodash'
import { useAppDispatch, useAppSelector } from '@store/hooks'
import { updateUserData, userSelectors } from '@store/user'
import { User } from '@store/user/types'

const updatingFieldsConfig = pick(fieldsConfig, [
    'login',
    'displayName',
    'firstName',
    'secondName',
    'email',
    'phone',
])

export const ProfileInfo: FC = () => {
    const [isEditMode, setIsEditMode] = useState(false)
    const userData = useAppSelector(userSelectors.selectData)
    const loadStatus = useAppSelector(userSelectors.selectStatus)
    const dispatch = useAppDispatch()

    const isLoading = loadStatus === 'loading'

    const toggleEditMode = useCallback((e?: MouseEvent<HTMLButtonElement>) => {
        e?.preventDefault()
        setIsEditMode(prevEditMode => !prevEditMode)
    }, [])

    const handleSaveProfile = useCallback(
        async (data: Omit<Partial<User>, 'id' | 'avatar'>) => {
            if (!isEditMode) {
                return
            }

            await dispatch(updateUserData(data))

            toggleEditMode()
        },
        [isEditMode, dispatch, toggleEditMode]
    )

    const defaultValues = useMemo(
        () => omit(userData, ['id', 'avatar']),
        [userData]
    )

    if (isLoading) {
        return <>Loading...</>
    }

    return (
        <>
            <Avatar avatar={userData?.avatarImage} />
            <Title className={styles.title}>{userData?.login ?? ''}</Title>
            <Form
                fields={updatingFieldsConfig}
                disabled={!isEditMode}
                defaultValues={defaultValues}
                SubmitButtonProps={{
                    children: isEditMode ? 'Сохранить' : 'Изменить',
                    variant: isEditMode ? 'contained' : 'outlined',
                    disabled: !isEditMode ? false : undefined,
                    onClick: !isEditMode ? toggleEditMode : undefined,
                }}
                CancelButtonProps={
                    isEditMode
                        ? {
                              children: 'Отменить',
                              variant: 'outlined',
                              onClick: toggleEditMode,
                          }
                        : undefined
                }
                onSubmit={handleSaveProfile}
            />
        </>
    )
}

export default ProfileInfo
