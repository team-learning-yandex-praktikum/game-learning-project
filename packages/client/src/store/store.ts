import { combineReducers, configureStore } from '@reduxjs/toolkit'
import { userReducer } from './user/userSlice'
import { settingsReducer } from './settings/settingsSlice'
import { sidebarReducer } from './sidebar/sidebarSlice'
import { gameReducer } from './game'
import { getWindow } from '@utils/document'
import { ssrReducer } from './ssr/ssrSlice'
import { leaderboardReducer } from './leaderboard/leaderboardSlice'

declare global {
    interface Window {
        initialState?: RootState
    }
}

export const reducer = combineReducers({
    user: userReducer,
    settings: settingsReducer,
    sidebar: sidebarReducer,
    game: gameReducer,
    ssr: ssrReducer,
    leaderboard: leaderboardReducer,
})

export const createStore = () => {
    const initialState = getWindow()?.initialState
    delete getWindow()?.initialState

    return configureStore({
        reducer: reducer,
        preloadedState: initialState,
    })
}

export const store = createStore()

export type RootState = ReturnType<typeof reducer>
export type AppDispatch = typeof store.dispatch
export type AppStore = typeof store
