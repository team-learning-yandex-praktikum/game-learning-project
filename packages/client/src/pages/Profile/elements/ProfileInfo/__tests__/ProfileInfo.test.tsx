import { setupServer } from 'msw/node'
import { render, screen } from '@testing-library/react'
import '@testing-library/jest-dom'
import { withProviders } from '../../../../../test-utils/render'
import {
    requestHandlers,
    userInfoStub,
} from '../../../../../test-utils/requestHandlers'
import ProfileInfo from '../ProfileInfo'
import { fieldsConfig, transformData } from '@utils'
import userEvent from '@testing-library/user-event'
import { store } from '@store'
import { userActions } from '@store/user'
import { LoadStatus } from '@utils/store/enums'

const server = setupServer(requestHandlers.updateProfile)

const fieldLabels = {
    login: fieldsConfig.login.label,
    displayName: fieldsConfig.displayName.label,
    firstName: fieldsConfig.firstName.label,
    secondName: fieldsConfig.secondName.label,
    email: fieldsConfig.email.label,
    phone: fieldsConfig.phone.label,
    changeButton: 'Изменить',
    cancel: 'Отменить',
    submit: 'Сохранить',
}

const getFields = () => {
    const login = screen.getByLabelText(fieldLabels.login)
    const displayName = screen.getByLabelText(fieldLabels.displayName)
    const firstName = screen.getByLabelText(fieldLabels.firstName)
    const secondName = screen.getByLabelText(fieldLabels.secondName)
    const email = screen.getByLabelText(fieldLabels.email)
    const phone = screen.getByLabelText(fieldLabels.phone)

    return {
        login,
        displayName,
        firstName,
        secondName,
        email,
        phone,
    }
}

describe('Profile Info', () => {
    beforeAll(() => server.listen())
    beforeEach(() => {
        store.dispatch(
            userActions.setData(transformData.from.dto(userInfoStub))
        )
    })
    afterEach(() => server.resetHandlers())
    afterAll(() => server.close())

    const testedComponent = withProviders(<ProfileInfo />)

    test('correctly rendered', () => {
        const { asFragment } = render(testedComponent)
        expect(asFragment()).toMatchSnapshot()
    })

    test('form buttons shouldnt be in the doc', async () => {
        render(testedComponent)

        const submitButton = screen.queryByText(fieldLabels.submit)
        const cancel = screen.queryByText(fieldLabels.cancel)

        expect(submitButton).not.toBeInTheDocument()
        expect(cancel).not.toBeInTheDocument()
    })

    test('fields should be readonly', async () => {
        render(testedComponent)

        const { login, displayName, firstName, secondName, email, phone } =
            getFields()

        expect(login).toBeDisabled()
        expect(displayName).toBeDisabled()
        expect(firstName).toBeDisabled()
        expect(secondName).toBeDisabled()
        expect(email).toBeDisabled()
        expect(phone).toBeDisabled()
    })

    test('fields should be enabled after click on change', async () => {
        render(testedComponent)

        const { login, displayName, firstName, secondName, email, phone } =
            getFields()
        const changeButton = screen.getByText(fieldLabels.changeButton)

        await userEvent.click(changeButton)

        expect(login).toBeEnabled()
        expect(displayName).toBeEnabled()
        expect(firstName).toBeEnabled()
        expect(secondName).toBeEnabled()
        expect(email).toBeEnabled()
        expect(phone).toBeEnabled()
    })

    test('cancel should reset form', async () => {
        render(testedComponent)

        const { displayName, secondName } = getFields()
        const changeButton = screen.getByText(fieldLabels.changeButton)

        await userEvent.click(changeButton)
        await userEvent.clear(displayName)
        await userEvent.type(displayName, 'ivan_ivanov')
        await userEvent.clear(secondName)
        await userEvent.type(secondName, 'Попов')

        const cancel = screen.getByText(fieldLabels.cancel)

        expect(displayName).toHaveValue('ivan_ivanov')
        expect(secondName).toHaveValue('Попов')

        await userEvent.click(cancel)

        expect(displayName).toHaveValue('')
        expect(secondName).toHaveValue(userInfoStub.second_name)
        expect(displayName).toBeDisabled()
        expect(changeButton).toBeInTheDocument()
    })

    test('submit should change state', async () => {
        render(testedComponent)

        const { displayName, secondName } = getFields()
        const changeButton = screen.getByText(fieldLabels.changeButton)

        await userEvent.click(changeButton)
        await userEvent.clear(displayName)
        await userEvent.type(displayName, 'ivan_ivanov')
        await userEvent.clear(secondName)
        await userEvent.type(secondName, 'Попов')

        const submit = screen.getByText(fieldLabels.submit)

        expect(submit).toBeEnabled()
        await userEvent.click(submit)

        const userState = store.getState().user

        expect(submit).not.toBeInTheDocument()
        expect(userState?.status).toBe(LoadStatus.complete)
        expect(userState?.data.displayName).toBe('ivan_ivanov')
        expect(userState?.data.secondName).toBe('Попов')
    })
})
