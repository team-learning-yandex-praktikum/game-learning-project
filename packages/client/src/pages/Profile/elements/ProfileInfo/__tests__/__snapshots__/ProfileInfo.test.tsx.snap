// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Profile Info correctly rendered 1`] = `
<DocumentFragment>
  <div
    class=""
  >
    <img
      alt="Avatar"
    />
    <input
      name="avatar"
      type="file"
    />
  </div>
  <h2
    class=""
  >
    ivan
  </h2>
  <div>
    <form
      id="r:id"
    >
      <div
        class=""
        data-error="false"
      >
        <div
          class=""
        >
          <label
            for="login"
          >
            Логин
          </label>
          <input
            disabled=""
            id="login"
            name="login"
            placeholder="ivan_ivanov"
            readonly=""
          />
        </div>
        <div />
      </div>
      <div
        class=""
        data-error="false"
      >
        <div
          class=""
        >
          <label
            for="displayName"
          >
            Никнейм
          </label>
          <input
            disabled=""
            id="displayName"
            name="displayName"
            placeholder="ivan"
            readonly=""
          />
        </div>
        <div />
      </div>
      <div
        class=""
        data-error="false"
      >
        <div
          class=""
        >
          <label
            for="firstName"
          >
            Имя
          </label>
          <input
            disabled=""
            id="firstName"
            name="firstName"
            placeholder="Иван"
            readonly=""
          />
        </div>
        <div />
      </div>
      <div
        class=""
        data-error="false"
      >
        <div
          class=""
        >
          <label
            for="secondName"
          >
            Фамилия
          </label>
          <input
            disabled=""
            id="secondName"
            name="secondName"
            placeholder="Иванов"
            readonly=""
          />
        </div>
        <div />
      </div>
      <div
        class=""
        data-error="false"
      >
        <div
          class=""
        >
          <label
            for="email"
          >
            Почта
          </label>
          <input
            disabled=""
            id="email"
            name="email"
            placeholder="ivan@test.ru"
            readonly=""
          />
        </div>
        <div />
      </div>
      <div
        class=""
        data-error="false"
      >
        <div
          class=""
        >
          <label
            for="phone"
          >
            Телефон
          </label>
          <input
            disabled=""
            id="phone"
            name="phone"
            placeholder="+7 999 999 99 99"
            readonly=""
          />
        </div>
        <div />
      </div>
    </form>
    <div>
      <button
        class=""
        form="r:id"
        type="submit"
      >
        Изменить
      </button>
    </div>
  </div>
</DocumentFragment>
`;
